File: chapter2/exercise_20/cpp/contains.cpp
#include <cassert>
#include <utility>
#include <vector>

template <class Container, class T> bool contains(const Container &c, const T &elem)
{
    return c.end() != std::ranges::find(c, elem);
}

template <class Container, class... T> bool contains_any(const Container &c, const T &&...elems)
{
    return (... || contains(c, elems));
}

template <class Container, class... T> bool contains_all(const Container &c, const T &&...elems)
{
    return (... && contains(c, elems));
}

template <class Container, class... T> bool contains_none(const Container &c, const T &&...elems)
{
    return !contains_any(c, std::forward<const T>(elems)...);
}

int main()
{
    std::vector<int> vec{1, 2, 33, 4, 5};
    assert(contains_any(vec, 9, 9, 1));
    assert(contains_all(vec, 1, 2, 33, 4, 5));
    assert(contains_none(vec, 9, 9, 9, 9));

    return 0;
}

File: chapter2/exercise_20/go/contains.go
package main

import "fmt"

func contains[T comparable](slice []T, value T) bool {
	for _, elem := range slice {
		if elem == value {
			return true
		}
	}
	return false
}

func contains_any[T comparable](slice []T, value ...T) bool {
	for _, elem := range value {
		if contains(slice, elem) {
			return true
		}
	}
	return false
}

func contains_all[T comparable](slice []T, value ...T) bool {
	for _, elem := range value {
		if !contains(slice, elem) {
			return false
		}
	}
	return true
}

func contains_none[T comparable](slice []T, value ...T) bool {
	for _, elem := range value {
		if contains(slice, elem) {
			return false
		}
	}
	return true
}

func main() {
	arr1 := []int{1, 2, 3, 4}
	fmt.Println(contains_all(arr1, 1, 2, 3, 4))
	fmt.Println(contains_all(arr1, 5, 2, 3, 4))
	fmt.Println(contains_any(arr1, 9, 9, 9, 1, 9))
	fmt.Println(contains_any(arr1, 9, 9, 9, 9, 9))
	fmt.Println(contains_none(arr1, 8, 8, 8, 8))
	fmt.Println(contains_none(arr1, 1, 8, 8, 8))

	arr2 := [4]int{1, 2, 3, 4}
	fmt.Println(contains_all(arr2[:], 1, 2, 3, 4))
	fmt.Println(contains_all(arr2[:], 5, 2, 3, 4))
	fmt.Println(contains_any(arr2[:], 9, 9, 9, 1, 9))
	fmt.Println(contains_any(arr2[:], 9, 9, 9, 9, 9))
	fmt.Println(contains_none(arr2[:], 8, 8, 8, 8))
	fmt.Println(contains_none(arr2[:], 1, 8, 8, 8))
}
